### PR TITLE
LUCENE-8671: Allow more fine-grained control over off-heap term dictionaries

### PR DESCRIPTION
**Pull request by s1monw**
This change allows to control if term-dictionaries are loaded off-heap or on
heap on a per reader basis. None NRT readers will access all term dictionaries
off heap including ID fields while readers that require fast ID access like all
readers used within index writer will by default only load non-ID like fields
heap. Additionally, IOContext has now the ability to specify if ram-usage should be
minimized and can control the off vs. on-heap decisions on a per-reader basis.

Off heap term dictionaries are still only used if the index input is memory mapped.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/store/ByteBuffersIndexInput.java [hunk 1]
```diff
@@ -193,5 +193,10 @@ private void ensureOpen() {
     if (in == null) {
       throw new AlreadyClosedException("Already closed.");
     }
-  }  
+  }
+
+  @Override
+  public boolean isMMapped() {
+    return true;
```
Comment (dweiss) [inline lucene/core/src/java/org/apache/lucene/store/ByteBuffersIndexInput.java]: Hi Simon. Whether this should return true depends on what byte buffers are used? The same applies to ByteBufferIndexInput, actually... I don't think you can generally tell whether the ByteBuffers the input operates on come from a mmap call or from somewhere else (even direct buffers don't have to be a result of mmap).

Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/store/ByteBuffersIndexInput.java]: it's rather mmapped or in ram so `true` is fine?

Comment (dweiss) [reply lucene/core/src/java/org/apache/lucene/store/ByteBuffersIndexInput.java]: Err... I didn't follow this change too closely. I thought the meaning of this is to distinguish between memory-mapped (off-heap) vs. heap-based storage?

Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/store/ByteBuffersIndexInput.java]: I will clarify the API.

Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/store/ByteBuffersIndexInput.java]: I removed this API went down a different route

**Comment by s1monw** [comment]
@mikemccand I agree - I close this one and try to break it into smaller chunks and different solutions. see #610

**Pull request closed without merging**